File: c++/newtons_method.cpp
#include <bits/stdc++.h>

#define MAX 99999

double func(double x){
    // Function to minimize, this can be changed to any function.
    // Parameters:
    //      :x: variable of function
    return x * x + 4 * cos(x);   
}


double dfunc(double x){
    // Derivative of minimized function
    return 2 * x - 4 * sin(x);
}


double ddfunc(double x){
    // Second derivative of minimized function
    return 2 - 4 * cos(x);
}

double newtons(double xPrev){
    // Implementation of newtons method 
    // Parameters:
    //      :xPrev: initial value of x
    std::clock_t start;
    start = std::clock();

    int iter = 0;
    double xNew, error = MAX;

    while (error > exp10(-5)){
        xNew = xPrev - dfunc(xPrev) / ddfunc(xPrev);
        error = fabs(xNew - xPrev);
        xPrev = xNew;
        iter++;
    }
    
    std::cout << "Newton's Method" << std::endl;
    std::cout << "Iteration : " << iter << std::endl;
    std::cout << "Computation Time : "  << std::clock() - start << std::endl;
    std::cout << "Minimum point of function : " << "x = " << xNew << ", f(x) = " << func(xNew) << std::endl; 

    return xNew;
}


int main(){


    newtons(1);

    return 0;
}

File: matlab/newtons_method.m
clear all
clc
tic
% Functions
func    = @(x) x ^ 2 + 4 * cos(x);   % function
d_func  = @(x) 2 * x - 4 * sin(x); % derivative of function
dd_func = @(x) 2 - 4 * cos(x);  % second derivative of function

% Newton's Method Algorithm
x_old = 1; 
iter = 0;
error = 100; % initial value for while loop
while (error > 10^-5) 
    
    x_new = x_old - d_func(x_old)/dd_func(x_old); % update x value
    error = abs(x_new - x_old); % tolarence
    
    x_old = x_new;   % update old x value
    iter = iter + 1; % iteration number
    
    
end
time = toc;
% Print results
fprintf("\tNewton's Method\n")
fprintf("Iteration : %d\n",iter);
fprintf("Computation time : %f second\n",time)
fprintf("Minimum point of functin : x = %f f(x) = %f\n",x_new,func(x_new));